### PR TITLE
Replace stats dict as Stats

### PR DESCRIPTION
**Pull request by PaulQbFeng**
Create class Stats that is initialized from a dict.

Can perform 
- stat1 + stat2
- stat1 - stat2
- stat.print_stats()
- stat.get(attribute, default)

### CODE REVIEW

File: stats.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        return print("\n".join([f"{k}: {v}" for k, v in self.__dict__.items()]))
+
+
+    def __add__(self, stats: StatsAnnot):
+        addition = dict()
+        for name, value in stats.__dict__.items():
+            if not hasattr(self, name):
+                addition[name] = value
+            else:
+                addition[name] = getattr(self, name) + value
```
Comment (Lysxia) [inline stats.py:28]: Where are the multiplicatively stacking stats?

Comment (PaulQbFeng) [reply stats.py:28]: here the mult stacking stats would be added normally e.g armor_reduction_perc (15% +25%)

Then the armor reductuon could be handled in a function during an attack to generate flat amount of armor to subtract

Comment (Lysxia) [reply stats.py:28]: My point is that you can't add `armor_reduction_perc`. 50% and 50% combine to 75%, not 100%.

Comment (PaulQbFeng) [reply stats.py:28]: Oh. I also need to check the behaviour for crit chance attack speed
other wise I guess a if statement on the mult stacking stat name to treat those cases should work

Comment (LinSun00) [reply stats.py:28]: I agree that it's probably wiser to make percent armor penetration stack multiplicatively. In New_structure_ideas, the add_stats of Stats has every correct stacking methods

Comment (PaulQbFeng) [reply stats.py:28]: 👌 I will copy that

File: stats.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    Champions, items, and rune shards can all be considered to have stats.
+    Stats in league all have a fixed way of being stacked, hence why it makes sense to define a class.
+    """
+
+    def __init__(self, stat_dict: Optional[StatsAnnot] = None):
+        if stat_dict is None:
+            stat_dict = {}
+            
+        for name, value  in stat_dict.items():
+            setattr(self, name, value)
```
Comment (michaelbenayoun) [inline stats.py:16]: I would rather store an underlying dictionary, and overight the `__getattr__` method to implement this mechanism.

File: stats.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    def print_stats(self):
+        return print("\n".join([f"{k}: {v}" for k, v in self.__dict__.items()]))
+
+
+    def __add__(self, stats: StatsAnnot):
+        addition = dict()
+        for name, value in stats.__dict__.items():
```
Comment (michaelbenayoun) [inline stats.py:24]: `stats` is already a dict no? `stats.items()`

File: stats.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            if not hasattr(self, name):
```
Comment (michaelbenayoun) [inline stats.py:25]:
```suggestion
            if name not in self._dict:
```
Where `self._dict` is the underlying dict mentioned above.

File: stats.py (hunk 1 above)
Comment (michaelbenayoun) [inline stats.py:28]:
```suggestion
                addition[name] = self._dict.get(name, 0) + value
```

Comment (michaelbenayoun) [reply stats.py:28]: This way you also don't have to do a branching

File: stats.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                addition[name] = value
+            else:
+                addition[name] = getattr(self, name) + value
+
+        return Stats(addition)
+
+    def __sub__(self, stats: StatsAnnot):
```
Comment (michaelbenayoun) [inline stats.py:32]: Same comments

File: stats.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                subtraction[name] = -value
+            else:
+                subtraction[name] = getattr(self, name) - value
+
+        return Stats(subtraction)
+
+
+    def get(self, attribute: str, default: float):
+        """Similar to a dict get with default value"""
+        return getattr(self, attribute, default)
```
Comment (michaelbenayoun) [inline stats.py:45]:
```suggestion
    def get(self, attribute: str, default: float):
        """Similar to a dict get with default value"""
        return self._dict.get(attribute, default)
```

File: stats_calculator.py [hunk 7]
```diff
@@ -19,14 +20,16 @@ def calculate_stat_from_level(base_stats: dict, stat_name: str, level: int):
     return calculate_flat_stat_from_level(stat, mean_growth_perlevel, level)
 
 
-def get_champion_base_stats(champion_stats, level):
+def get_champion_base_stats(champion_stats, level) -> StatsAnnot:
```
Comment (michaelbenayoun) [inline stats_calculator.py:23]:
```suggestion
def get_champion_base_stats(champion_stats, level) -> Stats:
```

File: stats_calculator.py [hunk 8]
```diff
@@ -19,14 +20,16 @@ def calculate_stat_from_level(base_stats: dict, stat_name: str, level: int):
...
 
-    return {stat_name: calculate_stat_from_level(champion_stats, stat_name, level) for stat_name in SCALING_STAT_NAMES}
+    return Stats(
+        {stat_name: calculate_stat_from_level(champion_stats, stat_name, level) for stat_name in SCALING_STAT_NAMES}
+    )
 
 
 # items
-def update_bonus_stat_with_item(total_item_stats, item_stats):
+def update_bonus_stat_with_item(total_item_stats: dict, item_stats: dict) -> None:
```
Comment (michaelbenayoun) [inline stats_calculator.py:32]: Not needed when it's `None` I think

File: stats_calculator.py [hunk 9]
```diff
@@ -35,11 +38,10 @@ def update_bonus_stat_with_item(total_item_stats, item_stats):
             total_item_stats[stat_name] += stat_value
 
 
-def get_items_total_stats(items):
+def get_items_total_stats(items: BaseItem) -> StatsAnnot:
```
Comment (michaelbenayoun) [inline stats_calculator.py:41]:
```suggestion
def get_items_total_stats(items: BaseItem) -> Stats:
```

File: stats.py (hunk 3 above)
Comment (PaulQbFeng) [reply stats.py:24]: stats would be a Stats object in this case (to do stat1 + stat2). Attribute wise addition / or setting if the attribute does not exist in stat1.

Comment (michaelbenayoun) [reply stats.py:24]: Ok, in this case the type annotation is wrong

File: stats_calculator.py (hunk 7 above)
Comment (PaulQbFeng) [reply stats_calculator.py:23]: btw what happens if i do `from stats import Stats` if I have a `Stats` annot and a `Stats` class ?

File: stats.py (hunk 3 above)
Comment (PaulQbFeng) [reply stats.py:24]: Indeed, I mixed things up

File: stats.py (hunk 2 above)
Comment (PaulQbFeng) [reply stats.py:16]: ok I see

File: tests/test_damage.py [hunk 10]
```diff
@@ -13,52 +13,52 @@ def tests_normal_auto_attack_cait_dummy_60():
     # Caitlyn level 1 with a doran's blade
     # No headshot no crit
     assert round(damage_auto_attack(base_attack_damage=62, bonus_attack_damage=13.4, lethality=0, attacker_level=1,
-                                    armor_pen_mult_factor=1, bonus_armor_pen_mult_factor=1, base_armor=dummy_60.orig_base_stats["armor"],
-                                    bonus_armor=dummy_60.orig_bonus_stats["armor"], damage_modifier_flat=0,
+                                    armor_pen_mult_factor=1, bonus_armor_pen_mult_factor=1, base_armor=dummy_60.orig_base_stats.armor,
```
[the remark this answers is not in the document]
Comment (PaulQbFeng) [reply tests/test_damage.py:16]: you mean changing `armor_pen_mult_factor` to `armor_pen_percent` in the function argument and calculate the coefficient inside the function ?

**Pull request merged**